packages/firestore/src/model/mutation.ts

```diff
@@ -576,23 +661,83 @@ export class TransformMutation extends Mutation {
     const transformResults = [] as FieldValue[];
     for (const fieldTransform of this.fieldTransforms) {
       const transform = fieldTransform.transform;
-      if (transform instanceof ServerTimestampTransform) {
-        let previousValue: FieldValue | null = null;
 
-        if (baseDoc instanceof Document) {
-          previousValue = baseDoc.field(fieldTransform.field) || null;
-        }
+      let previousValue: FieldValue = null;
+      if (baseDoc instanceof Document) {
+        previousValue = baseDoc.field(fieldTransform.field) || null;
+      }
 
-        transformResults.push(
-          new ServerTimestampValue(localWriteTime, previousValue)
+      let transformResult: FieldValue = null;
+      if (transform instanceof ServerTimestampTransform) {
+        transformResult = new ServerTimestampValue(
+          localWriteTime,
+          previousValue
+        );
+      } else if (transform instanceof ArrayUnionTransformOperation) {
+        transformResult = this.arrayUnionTransformResult(
+          transform,
+          previousValue
+        );
+      } else if (transform instanceof ArrayRemoveTransformOperation) {
+        transformResult = this.arrayRemoveTransformResult(
+          transform,
+          previousValue
         );
       } else {
         return fail('Encountered unknown transform: ' + transform);
       }
+
+      transformResults.push(transformResult);
     }
     return transformResults;
   }
 
+  /**
+   * Transforms the provided `previousValue` via the provided `arrayUnion`. Used
+   * for both local application and after server acknowledgement.
+   */
+  private arrayUnionTransformResult(
+    arrayUnion: ArrayUnionTransformOperation,
+    previousValue: FieldValue | null
+  ): FieldValue {
+    const result = this.coercedFieldValuesArray(previousValue);
+    for (const element of arrayUnion.elements) {
+      if (!array.includesEqualElement(result, element)) {
+        result.push(element);
+      }
+    }
+    return new ArrayValue(result);
+  }
+
+  /**
+   * Transforms the provided `previousValue` via the provided `arrayRemove`.
+   * Used for both local application and after server acknowledgement.
+   */
+  private arrayRemoveTransformResult(
+    arrayRemove: ArrayRemoveTransformOperation,
+    previousValue: FieldValue | null
+  ): FieldValue {
+    let result = this.coercedFieldValuesArray(previousValue);
+    for (const element of arrayRemove.elements) {
+      result = array.removeEqualElements(result, element);
+    }
+    return new ArrayValue(result);
+  }
+
+  /**
+   * Inspects the provided value, returning a mutable copy of the internal array
+   * if it's an ArrayValue and an empty mutable array if it's null or any other
+   * type of FieldValue.
+   */
+  private coercedFieldValuesArray(value: FieldValue | null): FieldValue[] {
+    if (value instanceof ArrayValue) {
+      return value.internalValue.slice();
+    } else {
+      // coerce to empty array.
```

Comment:
This seems like it is rarely what the user wants, and as such, I feel that we should be logging it at the very least.